src/WebJobs.Script/Host/BlobLeaseManager.cs

```diff
@@ -0,0 +1,173 @@
+        private const string HostContainerName = "azure-functions-host";
+
+        private readonly Timer _timer;
+        private readonly ICloudBlob _lockBlob;
+        private readonly TimeSpan _leaseTimeout;
+        private readonly TimeSpan _renewalInterval;
+        private readonly TimeSpan _leaseRetryInterval;
+        private readonly TraceWriter _traceWriter;
+        private readonly string _hostId;
+        private string _leaseId;
+        private bool _disposed = false;
+        
+        internal BlobLeaseManager(ICloudBlob lockBlob, TimeSpan leaseTimeout, string hostId, TraceWriter traceWriter)
+        {
+            _lockBlob = lockBlob;
+            _leaseTimeout = leaseTimeout;
+            _traceWriter = traceWriter;
+            _hostId = hostId;
+
+            // Renew the lease three seconds before it expires
+            _renewalInterval = leaseTimeout.Add(TimeSpan.FromSeconds(-3));
+
+            // Attempt to acquire a lease every 5 seconds
+            _leaseRetryInterval = TimeSpan.FromSeconds(5);
+
+            _timer = new Timer(o => AcquireOrRenewLeaseAsync().ContinueWith(t => { _traceWriter.Error($"Error processing host blob lease: {t.Exception}"); }, TaskContinuationOptions.OnlyOnFaulted),
+            null, TimeSpan.Zero, _leaseRetryInterval);
+        }
+
+        public event EventHandler HasLeaseChanged;
+
+        public bool HasLease => _leaseId != null;
+
+        public string LeaseId
+        {
+            get
+            {
+                return _leaseId;
+            }
+            private set
+            {
+                string previousId = _leaseId;
+                _leaseId = value;
+
+                if (string.Compare(previousId, _leaseId, StringComparison.OrdinalIgnoreCase) != 0)
+                {
+                    OnHasLeaseChanged();
+                }
+            }
+        }
+
+        private void OnHasLeaseChanged() => HasLeaseChanged?.Invoke(this, EventArgs.Empty);
+
+        public static async Task<BlobLeaseManager> CreateAsync(string accountConnectionString, TimeSpan leaseTimeout, string hostId, TraceWriter traceWriter)
+        {
+            if (leaseTimeout.TotalSeconds < 15 || leaseTimeout.TotalSeconds > 60)
+            {
+                throw new ArgumentOutOfRangeException(nameof(leaseTimeout), $"The {nameof(leaseTimeout)} should be between 15 and 60 seconds");
+            }
+
+            ICloudBlob blob = await GetLockBlobAsync(accountConnectionString);
+            var manager = new BlobLeaseManager(blob, leaseTimeout, hostId, traceWriter);
+            return manager;
+        }
+
+        public static BlobLeaseManager Create(string accountConnectionString, TimeSpan leaseTimeout, string hostId, TraceWriter traceWriter)
+        {
+            return CreateAsync(accountConnectionString, leaseTimeout, hostId, traceWriter).GetAwaiter().GetResult();
+        }
+
+        private async Task AcquireOrRenewLeaseAsync()
+        {
```

Comment:
What happens if the Contaner or Blob disappears out from under us? Should you recreate what we need if that happens?